### DISCUSSION TITLE
useFieldArray TS error: "Type string is not assignable to never"

### DISCUSSION THREAD

**Discussion by joostmeijles** [react-hook-form/react-hook-form#7586] [Q&A] [answered]
I get a "Type string is not assignable to never" Typescript error with the following application of `useForm` icw. `useFieldArray`:

```
type MyValues = {
  vals: number[];
};

function MyForm() {
  const { control } = useForm<MyValues>();
  useFieldArray({ control, name: 'vals' });
}
```

See also: https://stackblitz.com/edit/react-ts-hqumgu

What is the correct type usage for `useFieldArray`?

**Comment by Moshyfawn** [comment, marked as answer]
You can't do field arrays of non-object type, looking at `useFieldArray` `name` type; `FieldValues` is `Record<string, any>`
```ts
export declare const useFieldArray: <TFieldValues extends FieldValues = FieldValues, TFieldArrayName extends import("./types").ArrayPath<TFieldValues> = import("./types").ArrayPath<TFieldValues>, TKeyName extends string = "id">(props: UseFieldArrayProps<TFieldValues, TFieldArrayName, TKeyName>) => UseFieldArrayReturn<TFieldValues, TFieldArrayName, TKeyName>;
```

So, you probably should do
```ts
type MyValues = {
  vals: { value: number }[];
};
```

**Comment by eybel** [comment]
I have the same issue @Moshyfawn 

> type FormValues = {
>   products: {
>     title: String;
>     sub_title: String;
>     slug: String;
>     short_description: String;
>     long_description: String;
>     dimensions: String;
>     weight: String;
>     stock: String;
>     imageUrl: String;
>   }[];
> };
> 
> 
> const Admin = () => {
>   const [files, setFiles] = useState([]);
>   const { register, control, handleSubmit, formState: { errors }} = useForm<FormValues>();
> 
>   const { fields, remove, append } = useFieldArray<FormValues>(
>     {
>       control,
>       **name**: "variations",
>     }
>   );

I also tried this:

> type FormValues = {
     title: String;
     sub_title: String;
     slug: String;
     short_description: String;
     long_description: String;
     dimensions: String;
     weight: String;
     stock: String;
    imageUrl: String;
 };

'name' gets error saying: Type '"variations"' is not assignable to type '"products"'.ts. The expected type comes from property 'name' which is declared here on type 'UseFieldArrayProps<FormValues, "products", "id">'

I dont know what it means.

**Reply by Moshyfawn** [reply]
The `useFieldArray` **`name`** property should be a **key** of an **existing field** of type `array` of `object`.

Seeing your `FormValues` type, there's no field named "variations".

Basically, your `FormValues` type should include the **variations** `array` of type `object`. I'm guessing that variations is the sub-field of product. So, the type would be something like
```ts
products: {
  ... Your existing product types ...
  variations: {
    ... your variations type here ...
    // Should be an object, even if it's only, say, a `name` property
    // since RHF doesn't accept flat arrays (array of non-object types: `string`, `number`, etc.)
    // Example:
    name: "food" | "furniture"
    // or just
    name: string
  }[]
}[];
};
``` 

And your field array `name` would have to consist of **the parent field array name**, it's item's `index` and "variations". So, like
```js
const fieldArray = useFieldArray({
  name: `products.${productIndex}.variations`
})
```

So, you'll have to map your products field array fields and pass the `index` to a child component, where you gonna create your variations field array

**Reply by eybel** [reply]
@Moshyfawn  It worked! This is my final code. Is it ok how I did it?

```
type FormValues = {
  products: {
    title: String;
    sub_title: String;
    slug: String;
    short_description: String;
    long_description: String;
    dimensions: String;
    weight: String;
    stock: String;
    imageUrl: String;
  };
  variations: {
    option_name: String;
    option_sku: String;
    option_qty: String;
  }[];
};

const Admin = () => {
  const [files, setFiles] = useState([]);
  const {
    register,
    control,
    handleSubmit,
    formState: { errors },
  } = useForm<FormValues>();

  const { fields, append, remove } = useFieldArray<FormValues>({
    control,
    name: "variations",
  });
```

**Reply by Moshyfawn** [reply]
Yep, looks fine to me! ;)

**Reply by Moshyfawn** [reply]
You don't need to pass the `FormValues` to `useFieldArray` generic. It's type comes from the `control` prop

**Reply by eybel** [reply]
@Moshyfawn ok great! Alright, I'll remove it, those are the things that I cant understand yet with typescript, kind of new on that topic for me. Never cross my mind that a prop can bring the types also but I will follow your advice. Thank you!

**Reply by Moshyfawn** [reply]
No worries and great coding to ya! ;)

**Comment by eybel** [comment]
@Moshyfawn just wondering about something you said **_"since RHF doesn't accept flat arrays (array of non-object types: `string`, `number`, etc.)"_**

So if I pass this type as my "<FormType>", is wrong, right? :

```
short_description: string[];
```

In that case, I need to always make my type as follow? :

```
  short_description: {
    item: string;
  }[];
```

just wondering because I then had to update my Mongoose DB schema from type to object[] and also my react .map() functions to render each item of short_description...

Couldnt find info about this in the RHF docs

**Reply by Moshyfawn** [reply]
Yep, RHF needs to store the `id` field on the field's `object` to track the fields in the array. So, the field's shape should be of type `object`
```js
short_description: {
  item: string
}[]
```
is correct

**Reply by eybel** [reply]
Ok thank you for clarifying, that is why it was complaining so much when I set it up as an Array, I re-configured my DB schema, now I think everything works fine. Do you recommend any available course on React-Hook-Form? I really want to learn it, just started with it but I have to improve my skills with RHF
